### PR TITLE
Handle overwrite-if-exists for copy-without-render

### LINKED ISSUE #1029: Generate_context behavior when overwrite_if_exists = True and _copy_without_render is specified
* Cookiecutter version: 1.6.0
* Python version: py27, py35, py36
* Operating System: Ubuntu 16.04.3 LTS

### Description:

When overwrite_if_exists = True, and _copy_without_render is specified, the use of shutil.copytree in generate.py::generate_files means that the files cannot be over-written.  This seems like incorrect behavior to me.

### What I've run:

```
Traceback (most recent call last):
  File "/home/nicholasc/anaconda2/bin/cookiecutter", line 11, in <module>
    load_entry_point('cookiecutter==1.6.0', 'console_scripts', 'cookiecutter')()
  File "/home/nicholasc/anaconda2/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/nicholasc/anaconda2/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/nicholasc/anaconda2/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/nicholasc/anaconda2/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/nicholasc/anaconda2/lib/python2.7/site-packages/cookiecutter/cli.py", line 120, in main
    password=os.environ.get('COOKIECUTTER_REPO_PASSWORD')
  File "/home/nicholasc/anaconda2/lib/python2.7/site-packages/cookiecutter/main.py", line 94, in cookiecutter
    output_dir=output_dir
  File "/home/nicholasc/anaconda2/lib/python2.7/site-packages/cookiecutter/generate.py", line 328, in generate_files
    shutil.copytree(indir, outdir)
  File "/home/nicholasc/anaconda2/lib/python2.7/shutil.py", line 177, in copytree
    os.makedirs(dst)
  File "/home/nicholasc/anaconda2/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 17] File exists: '/home/nicholasc/tmp/ccdebug/projectname/docs/aibs_sphinx'
```

I have prepared a fix and will submit a PR that highlights the difference, and works around.

### PR DESCRIPTION
**Pull request by hugoduncan**
When overwrite_if_exists is specified, make _copy_without_render work
with existing target directories.

This is an alternative to #1030, which took a different approach,
removing the target directory before copying.  This PR does not remove
content when overwrite_if_exists is specified.

Replaces shutil.copytree with a version that accepts an exist_ok
keyword argument.  The exist_ok keyword argument is passed to
os.makedirs.  The function is based on shutils, with unused keyword
arguments removed.

Closes #1029

### CODE REVIEW

**Comment by donhector** [comment]
I'd love to see this PR merged!

_Here's my use case:_

My cookiecutter project contains a folder with many [Ansible](https://docs.ansible.com/ansible/latest/index.html) roles.

Ansible syntax itself uses Jinja2 templating (i.e. `{{ ansible-variable }}`) , so by default,  when `cookiecutter` is executed it thinks `{{ ansible-variable }}` is a Cookiecutter variable when it is not.

To avoid this issue, I've added my Ansible folder to `"_copy_without_render"` in my `cookiecutter.json` which works great.

However there's a big caveat: As I keep adding new roles to my Ansible folder, I'd like to re-run `cookiecutter` on already cut projects to pickup the newest roles, but when I do I get bitten by #1029

I know that instead of using `"_copy_without_render"` in my `cookiecutter.json` I can wrap all my Ansible files with `{% raw %}` and `{% endraw %}`, but this has several drawbacks:

- Wrapping the files Error prone
- Wrapping hundreds of Ansible files is unpleasant.
- Once I add the `raw` / `endraw` I loose YAML syntax validation in my code editor.

File: cookiecutter/utils.py [hunk 1]
```diff
@@ -110,3 +110,41 @@ def prompt_and_delete(path, no_input=False):
...
+    os.makedirs(dst, exist_ok=exist_ok)
+    errors = []
+    for name in names:
+        srcname = os.path.join(src, name)
+        dstname = os.path.join(dst, name)
+        try:
+            if os.path.islink(srcname):
+                linkto = os.readlink(srcname)
+                if os.path.isdir(srcname):
+                    copytree(srcname, dstname, symlinks)
```
Comment (donhector) [inline cookiecutter/utils.py]: Recursive calls to `copytree` should respect the `exist_ok` value that was provided:

`copytree(srcname, dstname, symlinks, exist_ok)`

File: cookiecutter/utils.py [hunk 2]
```diff
@@ -110,3 +110,41 @@ def prompt_and_delete(path, no_input=False):
...
+                else:
+                    shutil.copy2(srcname, dstname)
+            elif os.path.isdir(srcname):
+                copytree(srcname, dstname, symlinks)
```
Comment (donhector) [inline cookiecutter/utils.py]: Same applies here:

`copytree(srcname, dstname, symlinks, exist_ok)`

**Comment by kurtmckee** [comment]
I'm closing this PR due to its age.

**Pull request closed without merging**